### PR TITLE
elm @ Rakuten post

### PR DESCRIPTION
**Pull request by NemoOudeis**

### CODE REVIEW

File: content/post/elm-in-rakuten/index.md [hunk 1]
```diff
@@ -0,0 +1,1273 @@
+---
+title: "Elm in Rakuten"
+date: 2021-01-12T00:00:00+09:00
+hero: https://lucamug.github.io/post-Elm-in-Rakuten/images/elm-in-rakuten.png
```
Comment (jcayzac) [inline content/post/elm-in-rakuten/index.md:4]: The image should be hosted here IMHO.

File: content/post/elm-in-rakuten/index.md [hunk 2]
```diff
@@ -0,0 +1,1273 @@
...
+
+Elm was promising that all of our functions would be pure by design and that there would be **no errors at runtime**[^errors-at-runtime] anymore.
+
+[^errors-at-runtime]: In the front-end, **errors at runtime** are errors that happen in the browser. These errors can completely halt the functionality of the website and you, as the creator, may not even know about them because they are happening on other peoples' devices. Some tools allow you to get notified when these errors happen. These errors are usually JavaScript errors, for example, trying to access values that are null or undefined. 
+
+Moreover, **The Elm Architecture**, included in the Elm language, seemed an excellent way to structure an application. It ended up being one of the most influential Elm ideas in the front-end world, more about this later.
+
+![The Elm Architecture](https://lucamug.github.io/post-Elm-in-Rakuten/images/the-elm-architecture.jpg)
```
Comment (jcayzac) [inline content/post/elm-in-rakuten/index.md:44]: The image should be hosted here IMHO.

File: content/post/elm-in-rakuten/index.md [hunk 3]
```diff
@@ -0,0 +1,1273 @@
...
+
+[^semantic-versioning]: The Elm Package Manager allows you to check differences (functions added/removed/modified) in any published Elm package simply by running the command `elm diff` like in `elm diff elm/json 1.0.0 1.1.2`. Evan Czaplicki gives a live example of this feature in the video [Convergent Evolution](https://youtu.be/jl1tGiUiTtI).
+
+Languages with automatic type conversion, like JavaScript, can lead to surprises.[^funny-youtube-videos] What happens if we add a *number* and a *string* together? For example, `1` and `"2"`? Do we get `3`? Do we get `"12"`? Do we get an error? Do we get something else?
+
+Elm, in contrast, is strongly statically typed but types declaration are optional because the Elm compiler can infer 100% of them, including the types of all external libraries.
+
+[^funny-youtube-videos]: There are several funny videos about this topic on YouTube. One of the most popular is probably [What the... JavaScript?](https://youtu.be/2pL28CcEijU) by Kyle Simpsons
+
+![Dynamic Typing vs Static Typing](https://lucamug.github.io/post-Elm-in-Rakuten/images/dynamic-vs-static.png)
```
Comment (jcayzac) [inline content/post/elm-in-rakuten/index.md:193]: The image should be hosted here IMHO.

File: content/post/elm-in-rakuten/index.md [hunk 4]
```diff
@@ -0,0 +1,1273 @@
...
+
+
+
+
+
+
+
+## 3. “Making impossible states impossible”
+
+![Penrose Triangle](https://lucamug.github.io/post-Elm-in-Rakuten/images/penrose-triangle.jpg)
```
Comment (jcayzac) [inline content/post/elm-in-rakuten/index.md:239]: The image should be hosted here IMHO.

File: content/post/elm-in-rakuten/index.md [hunk 5]
```diff
@@ -0,0 +1,1273 @@
...
+*Dynamic Typing vs Static Typing, illustration by [Kolja Wilcke](https://twitter.com/01k), licensed under [CC BY 4.0](https://creativecommons.org/licenses/by/4.0/).*
+
+If you need to create a quick proof of concept, dynamic typing may be faster and the puzzle does resemble a giraffe even if it contains mistakes. But for robust applications and correct puzzle solutions, static typing is the right way to go.
+
+TypeScript, which adds optional static typing to JavaScript and is probably one of the best things to happen to JavaScript, can partially mitigate the issues with the JavaScript dynamic type system. But being a superset of JavaScript it needs to compromise on elegance and simplicity. For example, type declarations are optional (*any* as escape hatch), inference doesn’t cover all the code, it requires *type guards*[^type-guards], and not all JavaScript libraries have type annotations.
+
+
+
+[^type-guards]: [Type guards](https://www.typescriptlang.org/docs/handbook/advanced-types.html) are TypeScript expressions that perform a runtime check to discriminate between *custom types*. For example:{{< highlight typescript >}}function isFish(pet: Fish | Bird): pet is Fish {
```
Comment (jcayzac) [inline content/post/elm-in-rakuten/index.md:202]: Broken code fence

File: content/post/elm-in-rakuten/index.md [hunk 6]
```diff
@@ -0,0 +1,1273 @@
...
+
+
+
+
+
+
+
+
+
+
```
Comment (jcayzac) [inline content/post/elm-in-rakuten/index.md:235]: Lots of empty lines

File: content/post/elm-in-rakuten/index.md [hunk 7]
```diff
@@ -0,0 +1,1273 @@
...
+> “If it compiles, it works” **[^if-it-compiles-it-works]**
+    
+[^if-it-compiles-it-works]: This way of saying probably [originated in Haskell](https://wiki.haskell.org/Why_Haskell_just_works) but it applies to Elm too.
+
+Best practices should be automated as much as possible and the Elm compiler is playing an important role in this direction.
+
+The compiler guarantees that all edge cases are covered, something difficult to achieve with hand-made unit tests. Another advantage of the compiler static analysis is that it is extremely fast and can provide the exact location of errors.
+
+![John Carmack on Elm Errors](https://lucamug.github.io/post-Elm-in-Rakuten/images/john-carmack-on-errors.png)
```
Comment (jcayzac) [inline content/post/elm-in-rakuten/index.md:554]: The image should be hosted here IMHO.

File: content/post/elm-in-rakuten/index.md [hunk 8]
```diff
@@ -0,0 +1,1273 @@
...
+        
+[^unidirectional-data-flow]: The Elm Architecture is based on [**unidirectional data flow**](https://youtu.be/jl1tGiUiTtI?t=470) (a.k.a. **one-way data binding**) like React, in contrast to the **bidirectional data flow** (a.k.a. **two-way data binding**) of frameworks like Angular, Vue, and Svelte ([in Svelte two-way binding can be disabled](https://github.com/sveltejs/svelte/issues/54)). There have been issues with two-way data binding. For example, the many-to-many dependencies between the view and the model can create an infinite loop of cascading updates. Another issue is the lack of control of the change detection mechanism. It is an implicit behavior that is not easy to control. Unidirectional data flow tends to be more predictable.
+    
+![The Elm Architecture, animated](https://lucamug.github.io/post-Elm-in-Rakuten/images/the-elm-architecture.svg)
```
Comment (jcayzac) [inline content/post/elm-in-rakuten/index.md:632]: The image should be hosted here IMHO.

File: content/post/elm-in-rakuten/index.md [hunk 9]
```diff
@@ -0,0 +1,1273 @@
...
+    
+There are three building blocks in The Elm Architecture:
+    
+* `Model` --- the state of your application, the only thing that can mutate
+* `view` --- a way to turn your state into HTML
+* `update` --- a way to update your state based on the `Model` and the *messages*
+        
+If we zoom in on the `Elm` block in the diagram above, this is what is happening inside:        
+
+![The Elm Architecture, animated](https://lucamug.github.io/post-Elm-in-Rakuten/images/the-elm-architecture-animation.gif)
```
Comment (jcayzac) [inline content/post/elm-in-rakuten/index.md:645]: The image should be hosted here IMHO.
Also, it seems the same image is used multiple times, in different formats? (gif, png, svg…)

File: content/post/elm-in-rakuten/index.md [hunk 10]
```diff
@@ -0,0 +1,1273 @@
...
+
+## 12. The Elm debugger
+
+The built-in **Elm debugger**[^elm-debugger] is a useful tool to debug Elm applications. It shows the state of the application and keeps track of all the messages fired during the application's life. It also gives you the ability to go back in time.[^elm-debugger-example]    
+    
+[^elm-debugger]: More about the **Elm debugger** in [The Perfect Bug Report](https://elm-lang.org/news/the-perfect-bug-report)
+    
+[^elm-debugger-example]: The **Elm debugger** is usually disabled for applications that are released in production, but you can find an example of it in [elmjapan.org](https://elmjapan.org/) where it has been kept active for educational purposes.
+
+![The Elm Debugger](https://lucamug.github.io/post-Elm-in-Rakuten/images/debugger.gif)
```
Comment (jcayzac) [inline content/post/elm-in-rakuten/index.md:692]: The image should be hosted here IMHO.

File: content/post/elm-in-rakuten/index.md [hunk 11]
```diff
@@ -0,0 +1,1273 @@
...
+
+
+
+
+
+# What we don't like about Elm
+
+## 1. Not mainstream
+
+![Escape](https://lucamug.github.io/post-Elm-in-Rakuten/images/escape.jpg)
```
Comment (jcayzac) [inline content/post/elm-in-rakuten/index.md:1035]: The image should be hosted here IMHO.

File: content/post/elm-in-rakuten/index.md (hunk 5 above)
Comment (NemoOudeis) [reply content/post/elm-in-rakuten/index.md:145]: from what I can see fences with language break within footnotes, I guess luca resorted to the `highlight` shortcode 🤷‍♀️
I reformatted it a bit so it doesn't look so ugly (apparently hugo allows line breaks, just no blank lines)

**Review by boazy — APPROVED** [review]

**Review by lucamug — APPROVED** [review]
